Add GPU kernel for `tf.bincount`

yongtang opened the pull request:
This fix tries to address the issue raised in #11554 where there is no GPU support for `tf.bincount`.

This fix adds GPU support for `tf.bincount`.

This fix fixes #11554.

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang committed:
Split BincountOp with GPU and CPU version

This commit splits BincountOp with GPU and CPU version.
GPU implementation to follow.

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang committed:
Add GPU kernel for `tf.bincount`.

This fix tries to address the issue raised in 11554 where
there is no GPU support for `tf.bincount`.

This fix adds GPU support for `tf.bincount`.

This fix fixes 11554.

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang committed:
Update test cases for GPU support of `tf.bincount`

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang committed:
Address cases where input.size() == 0 or output.size() == 0

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

tensorflow-jenkins commented:
Can one of the admins verify this patch?

ekelsen commented:
Note that in the case of weights being passed to bincount, it is equivalent to unsorted_segment_sum.  I would prefer a solution that just replaces bincount with a call to unsorted_segment_sum in this case (ideally in the graph after the output size is determined).

For the case where no weights are given and they are assumed to be one, then it would be best to have the implementation call CUB as it will be _significantly_ faster.

vrv commented:
@ringw can you take a look at this?  Let me know if you want me to find someone else for the GPU kernel code review.

vrv commented:
Missed @ekelsen's comment, probably should do that :)

yongtang commented:
@ekelsen @vrv Thanks for the suggestion. Will spend some time and get `CUB` implementation added.

ekelsen commented:
@yongtang thanks for taking this on!  CUB for weights == 1 case will be great.  Avoiding code duplication by calling the gpu kernel that already exists for unsorted_segment_sum in the case where weight != None will also be very helpful!

yongtang committed:
Use CUB for histogram bincount when weight = 1.

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang committed:
Use unsorted_segment_sum when weights.size() != 0

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang committed:
Remove unneeded GPU kernels.

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang commented:
@ekelsen @vrv The PR has been updated with `CUB` used for `weights.size() == 0` and `unsorted_segment_sum` used for `weights.size() != 0`. Please take a look and let me know if there are any issues.

As `cub::DeviceHistogram::HistogramEven` depends on `atomicAdd` which does not have the `double` and `half` types, only `float` is enabled for `CUB` now. I will investigate further to see if there are ways to create wrapper for `double` and `half` versions of `atomicAdd`.

yongtang committed:
Update Bazel BUILD file with Buildifier

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

ekelsen commented on tensorflow/core/kernels/bincount_op.cc:
```diff
@@ -75,34 +67,86 @@ class BincountOp : public OpKernel {
+      auto output = output_t->flat<T>();
+      OP_REQUIRES_OK(
+          ctx, functor::BincountFunctor<Device, T>::Compute(ctx, arr, output));
+    } else {
```
Please handle the case of weights being None in the graph (ie edit the python code for the bincount op) by calling unsorted_segment_sum there after reduce_max.  The C++ code for bincount only needs to handle the case where the weights are implicitly one.

It's better to not have high level ops alias to the same underlying calculation; it only makes analyzing the graph harder.

yongtang replied:
Thanks @ekelsen. That part has been moved to python side.

ekelsen reviewed (commented):
Thanks for these changes, they look good.  I think we're almost there.  Thanks for the help!

ringw commented:
Sorry for the delay! Looks good pending @ekelsen's comments.

yongtang committed:
Move unsorted_segment_sum to python part.

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang committed:
Address review comments

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang commented on tensorflow/core/kernels/bincount_op.cc:
```diff
@@ -75,34 +67,86 @@ class BincountOp : public OpKernel {
+                          BincountOp<GPUDevice, type>)
 
-TF_CALL_NUMBER_TYPES(REGISTER);
+TF_CALL_float(REGISTER_KERNELS);
```
Thanks. Done. Though on my dev machine it seems only int32 and float works. Might be because `sm = 20`? I added int32 to the list. Don't know if additional registration works with a different dev machine or not.

yongtang commented on tensorflow/core/kernels/bincount_op_gpu.cu.cc:
```diff
@@ -0,0 +1,107 @@
+#define REGISTER_GPU_SPEC(type) \
+  template struct functor::BincountFunctor<GPUDevice, type>;
+
+TF_CALL_float(REGISTER_GPU_SPEC)
```
Done.

yongtang commented on tensorflow/python/kernel_tests/bincount_op_test.py:
```diff
@@ -57,33 +56,33 @@ def test_values(self):
       self.assertAllEqual(math_ops.bincount([1], maxlength=3).eval(), [0, 1])
       self.assertAllEqual(math_ops.bincount([], maxlength=3).eval(), [])
 
   def test_random_with_weights(self):
```
Thanks @ekelsen. `test_random_without_weights` has been added.

yongtang commented:
Thanks @ekelsen @ringw for the review. The PR has been updated. Please take a look and let me know if there area any issues.

ekelsen commented on tensorflow/core/kernels/bincount_op.cc:
```diff
@@ -75,34 +66,71 @@ class BincountOp : public OpKernel {
+
+    OP_REQUIRES(
+        ctx, (weights_t.NumElements() == 0),
+        errors::InvalidArgument("Weights should not be passed as it should be "
```
I'm sorry, I should've said this earlier, the CPU implementation needs to retain the ability to handle weights != None.  This is because of backwards compatibility - graphs that were created must still run correctly in a new version of TF.  (This is only true for the CPU version, as the GPU version didn't exist).

So basically, the CPU code needs to continue to be able to handle weights, but the graph code should stay as it is so that all _future_ graphs use unsorted_segment_sum instead.  Once six months have passed, we can remove the CPU code, although more likely, it will simply stay around forever.

yongtang replied:
Thanks @ekelsen. The CPU implementation has been added back for weight != None.

ekelsen commented on tensorflow/python/kernel_tests/bincount_op_test.py:
```diff
@@ -57,32 +56,44 @@ def test_values(self):
+    with self.test_session(use_gpu=True):
+      np.random.seed(42)
+      for dtype in [dtypes.int32, dtypes.int64, dtypes.float32, dtypes.float64]:
+        arr = np.random.randint(0, 1000, num_samples)
```
Right now dtype is unused, I think you want to switch the numpy function depending on the type (and only test int32 and float32 if using a gpu for the test).

yongtang replied:
Thanks @ekelsen. The tests have been updated.

yongtang committed:
Address review feedback.

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang commented:
Thanks @ekelsen for the review. The PR has been updated. Please take a look.

ekelsen reviewed (commented):
One final request - can you provide some benchmarks on both CPU and GPU of the new version?

Then I think we're ready.  Thanks so much!

yongtang committed:
Add benchmark tests.

Some run result:
```
Running main() from test_main.cc
Benchmark                  Time(ns) Iterations
----------------------------------------------
BM_Bincount_cpu_32_1000      114922       5150   285.1M items/s
BM_Bincount_cpu_32_2000      124291       5524   263.6M items/s
BM_Bincount_cpu_32_5000      159548       4287   205.4M items/s
BM_Bincount_cpu_64_1000      145006       4793   452.0M items/s
BM_Bincount_cpu_64_2000      150301       4457   436.0M items/s
BM_Bincount_cpu_64_5000      180001       3880   364.1M items/s
BM_Bincount_cpu_128_1000     204993       3405   639.4M items/s
BM_Bincount_cpu_128_2000     209144       3311   626.7M items/s
BM_Bincount_cpu_128_5000     231580       3003   566.0M items/s

BM_Bincount_gpu_32_1000       61178      10000   535.6M items/s
BM_Bincount_gpu_32_2000       61021      10000   537.0M items/s
BM_Bincount_gpu_32_5000       61177      10000   535.6M items/s
BM_Bincount_gpu_64_1000       61317      10000   1068.8M items/s
BM_Bincount_gpu_64_2000       60726      10000   1079.2M items/s
BM_Bincount_gpu_64_5000       61721      10000   1061.8M items/s
BM_Bincount_gpu_128_1000      69935      10000   1874.2M items/s
BM_Bincount_gpu_128_2000      79760       9852   1643.3M items/s
BM_Bincount_gpu_128_5000     100407       6974   1305.4M items/s
```

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang commented:
Thanks @ekelsen. The PR has been updated with tests added. Here are some numbers for the new version:
```
Running main() from test_main.cc
Benchmark                  Time(ns) Iterations
----------------------------------------------
BM_Bincount_cpu_32_1000      114922       5150   285.1M items/s
BM_Bincount_cpu_32_2000      124291       5524   263.6M items/s
BM_Bincount_cpu_32_5000      159548       4287   205.4M items/s
BM_Bincount_cpu_64_1000      145006       4793   452.0M items/s
BM_Bincount_cpu_64_2000      150301       4457   436.0M items/s
BM_Bincount_cpu_64_5000      180001       3880   364.1M items/s
BM_Bincount_cpu_128_1000     204993       3405   639.4M items/s
BM_Bincount_cpu_128_2000     209144       3311   626.7M items/s
BM_Bincount_cpu_128_5000     231580       3003   566.0M items/s

BM_Bincount_gpu_32_1000       61178      10000   535.6M items/s
BM_Bincount_gpu_32_2000       61021      10000   537.0M items/s
BM_Bincount_gpu_32_5000       61177      10000   535.6M items/s
BM_Bincount_gpu_64_1000       61317      10000   1068.8M items/s
BM_Bincount_gpu_64_2000       60726      10000   1079.2M items/s
BM_Bincount_gpu_64_5000       61721      10000   1061.8M items/s
BM_Bincount_gpu_128_1000      69935      10000   1874.2M items/s
BM_Bincount_gpu_128_2000      79760       9852   1643.3M items/s
BM_Bincount_gpu_128_5000     100407       6974   1305.4M items/s

```

vrv commented:
@tensorflow-jenkins test this please

@ekelsen can you take one final look?

ekelsen reviewed (approved):
LGTM!

ringw commented:
LGTM

yongtang commented:
I couldn't see all of the CI failure logs, though one failure is:
```
ERROR: /tmpfs/src/github/tensorflow/tensorflow/examples/wav_to_spectrogram/BUILD:40:1: Couldn't build file tensorflow/examples/wav_to_spectrogram/wav_to_spectrogram_test: Linking of rule '//tensorflow/examples/wav_to_spectrogram:wav_to_spectrogram_test' failed (Exit 34). Note: Remote connection/protocol failed with: connection failed: com.google.devtools.build.lib.remote.RetryException: after 6 attempts: io.grpc.StatusRuntimeException: UNAVAILABLE: Subchannel shutdown invoked
```

I think that is an internal CI or network failure?

ringw commented:
It sure looks like a flaky network failure.

@tensorflow-jenkins please re-test

ringw commented:
@tensorflow-jenkins test this please

yongtang commented:
Half of the Jenkins tests passed. For the other half they are not invoked. I think a label of `kokoro:force-run` has to be attached to this PR to invoke the other tests to run.

yongtang commented:
Most of the test passed now. The error seems to be related to the network failure:
```
WARNING: /tmpfs/src/github/tensorflow/tensorflow/core/BUILD:1786:1: in includes attribute of cc_library rule //tensorflow/core:framework_headers_lib: '../../external/nsync/public' resolves to 'external/nsync/public' not below the relative path of its package 'tensorflow/core'. This will be an error in the future. Since this rule was created by the macro 'cc_header_only_library', the error might have been caused by the macro implementation in /tmpfs/src/github/tensorflow/tensorflow/tensorflow.bzl:1048:30
ERROR: /tmpfs/tmp/bazel/external/boringssl/BUILD:115:1: Couldn't build file external/boringssl/_objs/crypto/external/boringssl/src/crypto/mem.o: C++ compilation of rule '@boringssl//:crypto' failed (Exit 34). Note: Remote connection/protocol failed with: connection failed: com.google.devtools.build.lib.remote.RetryException: after 6 attempts: io.grpc.StatusRuntimeException: UNAVAILABLE: Subchannel shutdown invoked
	at com.google.devtools.build.lib.remote.Retrier.execute(Retrier.java:217)
	at com.google.devtools.build.lib.remote.GrpcRemoteExecutor.lambda$executeRemotely$2(GrpcRemoteExecutor.java:136)
	at com.google.devtools.build.lib.remote.Retrier.execute(Retrier.java:203)
	at com.google.devtools.build.lib.remote.GrpcRemoteExecutor.executeRemotely(GrpcRemoteExecutor.java:126)
```

I think a re-run may resolve the issue?

gunan commented:
Looks like boncount_op_test is failing to build on multiple configurations, and we seem to be changing the public TF API with this change. 
Could you take a look at the failures?

yongtang commented:
Thanks @gunan for the help. Let me take a look and fix it.

yongtang committed:
Remove Bincount from `hidden_ops.txt`, and remove unneeded

libraries so that Jenkins could pass.

Bincount should not be added `hidden_ops.txt` as it will cause
compatibility test fail.

And the following libs should not be needed in CI/CD:
```
-        "@local_config_cuda//cuda:cublas",
-        "@local_config_cuda//cuda:cuda_driver",
-        "@local_config_cuda//cuda:cudnn",
-        "@local_config_cuda//cuda:cufft",
-        "@local_config_cuda//cuda:curand",
```

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang commented:
@gunan The api compatibility test failure is caused by addition of `Bincount` into `hidden.txt`. I have removed it from `hidden.txt`.

For the gpu test linkage failure, on my dev machine previously I added the following so that it could pass locally:
```
        "@local_config_cuda//cuda:cublas",
        "@local_config_cuda//cuda:cuda_driver",
        "@local_config_cuda//cuda:cudnn",
        "@local_config_cuda//cuda:cufft",
        "@local_config_cuda//cuda:curand",
```

It seems that on CI/CD those libs should not be specified explicitly. I have removed those extra dependencies from the bazel `BUILD` build file. I think that will resolve the issue on Jenkins CI/CD.

yongtang commented:
@gunan The PR has been updated. Can you start the Jenkins build again? Thanks for the help and patient during the process and really appreciate that.

gunan commented:
Thank you for the updates.
Jenkins, test this please.
